libs/ml/include/ml/ops/activations/logsoftmax.hpp

```diff
@@ -64,17 +64,25 @@ class LogSoftmax : public fetch::ml::Ops<T>
 
     // return_signal.InlineMultiply(t);
 
-    // 1D softmax
-    if (inputs.front()->shape().size() == 1)
+    // 1D softmax with 1 batch dimension
+    if (inputs.front()->shape().size() == 2)
     {
-      typename ArrayType::Type sum = return_signal.Sum();
+      assert(axis_ == 1);
+      ArrayType sum = ReduceSum(return_signal, 0);
+
       t.InlineMultiply(sum);
     }
-    // 2D softmax
-    else if (inputs.front()->shape().size() == 2)
+    // 2D softmax with 1 batch dimension
+    else if (inputs.front()->shape().size() == 3)
     {
-      ArrayType sum;
-      sum = ReduceSum(return_signal, 1 - axis_);
+      assert((axis_ == 1) || (axis_ == 0));
+      ArrayType sum = return_signal.Slice(0, 1 - axis_).Copy();
+      for (size_t i = 0; i < return_signal.shape()[2]; i++)
+      {
+        auto cur_sum = ReduceSum(return_signal.Slice(i, 2).Copy().Squeeze(), 1 - axis_).View();
```

Comment:
i might be mistaken, but isn't this part:
return_signal.Slice(i, 2).Copy().Squeeze()
effectively a more expensive version of this:
return_signal.View(i).Copy()